File: linux/src/meeting_recorder/core/app_info.py
"""
Application identity for the About dialog: description, project links, authorship,
and a best-effort runtime version resolver.

The version is not baked into the source tree — releases substitute ``@VERSION@``
into the packaging metadata at build time (see ``linux/packaging/``), so the only
authoritative version at runtime is what pacman recorded when the app was
installed. ``resolve_version()`` queries that and returns ``None`` when it can't
be determined (e.g. running from a source checkout), in which case the About
dialog simply omits the version line.

The querying is a thin GTK-free wrapper; the parsing is pure and unit-tested via an
injectable ``run_fn`` (mirrors the ``services.system_installer`` test-seam pattern).
"""

from __future__ import annotations

import logging
import subprocess
from collections.abc import Callable

from ..config.defaults import APP_ID, APP_NAME

logger = logging.getLogger(__name__)

__all__ = [
    "APP_ID",
    "APP_NAME",
    "DESCRIPTION",
    "REPOSITORY",
    "ISSUE_URL",
    "DEVELOPER_NAME",
    "DEVELOPERS",
    "COPYRIGHT",
    "PACKAGE_NAME",
    "resolve_version",
]

DESCRIPTION = "Records meetings and generates transcripts and structured notes using AI."
REPOSITORY = "https://github.com/jmarceno/gravaai"
ISSUE_URL = f"{REPOSITORY}/issues"
DEVELOPER_NAME = "jmarceno"
DEVELOPERS = ["jmarceno <https://github.com/jmarceno>"]
COPYRIGHT = "© 2026 jmarceno (hard fork of Dipak Yadav's meeting-recorder)"

# The name the Arch package is installed under (see linux/packaging/).
PACKAGE_NAME = "meeting-recorder"

# A runner returns a command's stdout, or None if the command is missing / fails.
RunFn = Callable[[list[str]], str | None]


def _default_run(argv: list[str]) -> str | None:
    """Run ``argv`` and return stripped stdout, or None if it isn't available/fails."""
    try:
        result = subprocess.run(
            argv,
            capture_output=True,
            text=True,
            timeout=5,
        )
    except (OSError, subprocess.SubprocessError) as exc:
        logger.debug("version query %s failed: %s", argv[0], exc)
        return None
    if result.returncode != 0:
        return None
    return result.stdout


def _parse_pacman(stdout: str) -> str | None:
    """``pacman -Q pkg`` prints ``<name> <version>``; take the version field."""
    parts = stdout.split()
    if len(parts) >= 2:
        return parts[1]
    return None


# Single probe: pacman is the only supported package manager (Arch-only fork).
_VERSION_QUERIES: list[tuple[list[str], Callable[[str], str | None]]] = [
    (["pacman", "-Q", PACKAGE_NAME], _parse_pacman),
]


def resolve_version(run_fn: RunFn | None = None) -> str | None:
    """Return the installed package version, or None if it can't be determined.

    Queries pacman; returns None when pacman doesn't know the package — e.g. a
    source checkout — so callers can gracefully omit the version.
    """
    run = run_fn or _default_run
    for argv, parse in _VERSION_QUERIES:
        stdout = run(argv)
        if stdout is None:
            continue
        version = parse(stdout)
        if version:
            return version
    return None

File: linux/src/core/app_info.rs
//! App identity + installed-version resolution.

use std::path::PathBuf;

pub const DESCRIPTION: &str =
    "Records meetings and generates transcripts and structured notes using AI.";
pub const REPOSITORY: &str = "https://github.com/jmarceno/gravaai";
pub const ISSUE_URL: &str = "https://github.com/jmarceno/gravaai/issues";
pub const DEVELOPER_NAME: &str = "jmarceno";
pub const DEVELOPERS: &[&str] = &["jmarceno <https://github.com/jmarceno>"];
pub const COPYRIGHT: &str = "© 2026 jmarceno (hard fork of Dipak Yadav's meeting-recorder)";
pub const PACKAGE_NAME: &str = "meeting-recorder";

/// Read the installed pacman package version (`pacman -Q meeting-recorder`
/// prints `<name> <version>`). Returns None on a source checkout or when
/// pacman is unavailable. The command runner is injectable for tests.
pub fn resolve_version(run: &dyn Fn() -> Option<String>) -> Option<String> {
    let out = run()?;
    let mut parts = out.split_whitespace();
    let name = parts.next()?;
    let version = parts.next()?;
    // Must be exactly our package (pacman errors like "error: package ..."
    // have extra tokens and a foreign name).
    if name != PACKAGE_NAME || parts.next().is_some() {
        return None;
    }
    if version.trim().is_empty() {
        return None;
    }
    Some(version.to_string())
}

/// Version stamped into the AppImage at pack time
/// (`usr/share/meeting-recorder/VERSION`). Only consulted when we are
/// actually running from our own AppImage (host IDE AppImages are ignored).
fn version_from_appimage() -> Option<String> {
    let appdir = std::env::var_os("APPDIR").map(PathBuf::from)?;
    crate::utils::exe::own_appimage()?;
    let text = std::fs::read_to_string(appdir.join("usr/share/meeting-recorder/VERSION")).ok()?;
    let version = text.trim();
    if version.is_empty() {
        None
    } else {
        Some(version.to_string())
    }
}

pub fn installed_version() -> Option<String> {
    if let Some(v) = version_from_appimage() {
        return Some(v);
    }
    resolve_version(&|| {
        std::process::Command::new("pacman")
            .args(["-Q", PACKAGE_NAME])
            .output()
            .ok()
            .filter(|o| o.status.success())
            .map(|o| String::from_utf8_lossy(&o.stdout).into_owned())
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_pacman_output() {
        let v = resolve_version(&|| Some("meeting-recorder 1.2.0-1\n".to_string()));
        assert_eq!(v.as_deref(), Some("1.2.0-1"));
    }

    #[test]
    fn tolerates_garbage() {
        assert_eq!(resolve_version(&|| None), None);
        assert_eq!(resolve_version(&|| Some("".to_string())), None);
        assert_eq!(resolve_version(&|| Some("   \n".to_string())), None);
        assert_eq!(
            resolve_version(&|| Some("error: package not found\n".to_string())),
            None
        );
    }
}
